### PR TITLE
Add power set notation

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

The previous implementation in #7329 have some limitations that it can only be applicable for finite sets, and generally, computation can be slow even for a handful of elements.
I've added a full symbolic representation to resolve such issues.

I also think that using powerset can resolve failings like `FiniteSet(1, 2) not in FiniteSet(S.EmptySet)` because it can use subset testing for membership testing.

#### Other comments

I've also fixed the rst section ordering, added missing docs for some other set classes.

The doc for the PowerSet would look like
image

#### Release Notes

- sets
  - Added `PowerSet` for representing symbolic power set.
  - Fixed table-of-contents in the set documentation.
  - Added missing documentation about `SymmetricDifference`, `ConditionSet`.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
I wonder if this should default to evaluate=False and only expand a FiniteSet on doit. Expanding even a FiniteSet of size 10 could be very expensive...

**Comment by oscarbenjamin** [comment]
This looks good though!

**Comment by asmeurer** [comment]
@sylee957 I like the way you show what the rendered docs look like. I wonder if this is something we can automate somehow, perhaps using @sympy-bot or GitHub actions (I've put SymPy on the waiting list).

**Comment by certik** [comment]
@sylee957 thanks for your contribution.

@asmeurer yes, I think we should have some way to see the docs from a PR, whether a CI should simply upload the docs into some branch somewhere and host it over html, or if there is some better way, that I don't know.

File: sympy/sets/powerset.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        return super(PowerSet, cls).__new__(cls, arg)
+
+    @property
+    def arg(self):
+        return self.args[0]
+
+    @_sympifyit('other', NotImplemented)
+    def _contains(self, other):
+        if not isinstance(other, Set):
+            raise ValueError('{} must be a set.'.format(other))
```
Comment (oscarbenjamin) [inline sympy/sets/powerset.py]: This line isn't tested.

File: sympy/sets/powerset.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        arg = self.arg
+        ret = fuzzy_bool(arg.is_superset(other))
+        if ret is not None:
+            return ret
+        return None
```
Comment (oscarbenjamin) [inline sympy/sets/powerset.py]: This line also isn't tested. Perhaps:
```julia
In [10]: PowerSet({x}).contains({y})                                                                                              
Out[10]: {y} ∈ {∅, {x}}

In [11]: PowerSet({x}).contains({y}).subs(x, y)                                                                                   
Out[11]: True
```

File: sympy/sets/powerset.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/sets/powerset.py]: This isn't consistent with other sets:
```julia
In [7]: PowerSet({1})._contains(2)                                                                                                

In [8]: PowerSet({1}, evaluate=False)._contains(2)                                                                                
---------------------------------------------------------------------------
ValueError
```

Comment (oscarbenjamin) [reply sympy/sets/powerset.py]: `_contains(2)` should evaluate to False. Looks like EmptySet needs `_eval_Eq`:
```julia
In [7]: P = PowerSet({1})                                                                                                         

In [8]: P                                                                                                                         
Out[8]: {∅, {1}}

In [9]: print(P._contains(2))                                                                                                     
None

In [10]: Eq(S.EmptySet, 2)                                                                                                        
Out[10]: ∅ = 2
```

Comment (sylee957) [reply sympy/sets/powerset.py]: I don't think that the relations like below are well defined across various set theories. An element who cannot be a set would be called urelement, but it is not shared across every axioms.

image

image

image

But only these statements may be determined to be false

image

image

**Comment by sylee957** [comment]
> I wonder if this should default to evaluate=False and only expand a FiniteSet on doit. Expanding even a FiniteSet of size 10 could be very expensive...

I think that every `evaluate` should follow `global_evaluate[0]` or the context manager `with evaluate(False)` may not behave properly.
Instead, I'm thinking of either adding `doit`, or a rewriting method

**Comment by oscarbenjamin** [comment]
Not evaluating by default and having a rewrite method probably makes the most sense.

In general though I think that global_evaluate is poorly designed since:
1. Some objects default to evaluate=True (e.g. Pow) and others to evaluate=False (e.g. Derivative).
2. If all objects respect `global_evaluate` then there's no way for them to have different defaults.
3. The default value for `global_evaluate` should be `None` rather than True/False so that objects know when to use their default.
4. Explicit `evaluate=` arguments provided when creating an object should override `global_evaluate`.

This discussion is not related to this issue there and probably belongs at #17280

**Comment by oscarbenjamin** [comment]
The doctests failed:
```
File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/sets/powerset.py", line 45, in sympy.sets.powerset.PowerSet
Failed example:
    PowerSet(FiniteSet(1, 2, 3))
Expected:
    {EmptySet(), {1}, {2}, {3}, {1, 2}, {1, 3}, {2, 3}, {1, 2, 3}}
Got:
    PowerSet({1, 2, 3})
**********************************************************************
File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/sets/powerset.py", line 50, in sympy.sets.powerset.PowerSet
Failed example:
    PowerSet(S.EmptySet)
Expected:
    {EmptySet()}
Got:
    PowerSet(EmptySet())
**********************************************************************
File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/sets/powerset.py", line 52, in sympy.sets.powerset.PowerSet
Failed example:
    PowerSet(PowerSet(S.EmptySet))
Expected:
    {EmptySet(), {EmptySet()}}
Got:
    PowerSet(PowerSet(EmptySet()))
```
Is that intentional so that the doctests just need to be updated?

**Comment by sylee957** [comment]
I’ve decided to change powerset unevaluated by default, so the doctest may fail.
I think that some tests may also have to be added for iterators and subsets

File: sympy/sets/powerset.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    @_sympifyit('other', NotImplemented)
+    def _contains(self, other):
+        if not isinstance(other, Set):
+            return None
+
+        arg = self.arg
+        ret = fuzzy_bool(arg.is_superset(other))
+        if ret is not None:
+            return ret
+        return None
```
Comment (oscarbenjamin) [inline sympy/sets/powerset.py]: Can this just be `return self.arg.is_superset(other)`? Does `is_superset` already return a fuzzy bool?

File: sympy/sets/powerset.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        arg = _sympify(arg)
+
+        if not isinstance(arg, Set):
+            raise ValueError('{} must be a set.'.format(arg))
+
+        return super(PowerSet, cls).__new__(cls, arg)
+
+    @property
+    def arg(self):
+        return self.args[0]
```
Comment (oscarbenjamin) [inline sympy/sets/powerset.py]: I think that `arg` can have a better name. Perhaps `base_set`?

File: sympy/sets/powerset.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        ret = fuzzy_bool(arg.is_superset(other))
+        if ret is not None:
+            return ret
+        return None
+
+    def _eval_is_subset(self, other):
+        if isinstance(other, PowerSet):
+            ret = self.arg.is_subset(other.arg)
+            if ret is not None:
+                return ret
```
Comment (oscarbenjamin) [inline sympy/sets/powerset.py]: `return self.arg.is_subset...`

File: sympy/sets/sets.py [hunk 6]
```diff
@@ -366,6 +366,12 @@ def is_subset(self, other):
 
         """
         if isinstance(other, Set):
+            dispatch = getattr(self, '_eval_is_subset', None)
+            if dispatch is not None:
+                ret = dispatch(other)
+                if ret is not None:
+                    return ret
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py]: Can we just `return dispatch(other)`?

Comment (sylee957) [reply sympy/sets/sets.py]: If it returns, then the check using the computation of the intersection may not be true.
Though generally, I would think that having to compute intersection every time and relying on the automatic canonicalization may not be a good idea for testing the subset.

**Comment by oscarbenjamin** [comment]
Looks good. Merging.

**Pull request merged**